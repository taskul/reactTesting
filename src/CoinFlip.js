import { useState } from "react";
import heads from './heads.png';
import tails from './tails.png';

const CoinFlip = (props) => {
    const [coin, setCoin] = useState(heads);
    const [coinCaption, setCoinCaption] = useState('heads');
    const [headCount, setHeadCount] = useState(0);
    const [tailCount, setTailCount] = useState(0);
    const [totalFlilps, setTotalFlips] = useState(0);

    const choice = (arr) => {
        const idx = Math.floor(Math.random() * arr.length)
        return arr[idx]
    }
    const flipCoin = () => {
        const coin = choice(props.coin)
        setTotalFlips(totalFlilps + 1)
        if (coin.caption === 'heads') {
            setHeadCount(headCount + 1)
            setCoinCaption('heads')
        } else {
            setTailCount(tailCount + 1)
            setCoinCaption('tails')
        }
        setCoin(coin.src)
    }
    return (
        <div>
            <img src={coin} onClick={flipCoin} alt={coinCaption} data-testid="coin" />
            <p>Out of {totalFlilps}, there have been {headCount} heads and {tailCount} tails.</p>
        </div>
    )
}

CoinFlip.defaultProps = {
    coin: [
        {
            src: heads,
            caption: 'heads'
        },
        {
            src: tails,
            caption: 'tails'
        }
    ]
}

export default CoinFlip;