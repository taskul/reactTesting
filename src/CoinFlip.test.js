import { render, fireEvent } from "@testing-library/react";
import CoinFlip from "./CoinFlip";


beforeEach(function () {
    jest
        .spyOn(Math, "random")
        .mockReturnValueOnce(0.25)
        .mockReturnValueOnce(0.75);
});

afterEach(function () {
    Math.random.mockRestore();
});


// smoke test
it('renders without crashing', () => {
    render(<CoinFlip />)
})

// snapshot test
it('it should match snapshot', () => {
    const { asFragment } = render(<CoinFlip />);
    expect(asFragment()).toMatchSnapshot();
})

it('should flip a coin when clicked', () => {
    const { queryByAltText, queryByTestId, debug } = render(<CoinFlip />)

    expect(queryByAltText('heads')).toBeInTheDocument();

    const coin = queryByTestId('coin')
    fireEvent.click(coin)

    expect(queryByAltText('heads')).toBeInTheDocument();

    fireEvent.click(coin)

    expect(queryByAltText('tails')).toBeInTheDocument();
})