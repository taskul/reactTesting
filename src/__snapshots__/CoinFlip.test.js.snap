// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <img
      alt="heads"
      data-testid="coin"
      src="heads.png"
    />
    <p>
      Out of 0, there have been 0 heads and 0 tails.
    </p>
  </div>
</DocumentFragment>
`;
